109-custom-client-authentication.md

```diff
@@ -0,0 +1,156 @@
+
+Strimzi users typically integrate with these services in two different situations:
+* Some Strimzi users use only some parts of the Strimzi project.
+  They use managed Apache Kafka services to run the Kafka cluster (brokers, controllers, ...).
+  And they combine it with a Kafka Connect cluster or Strimzi HTTP Bridge running in a self-managed way using Strimzi.
+* Other Strimzi users use Mirror Maker 2 to integrate between different Apache Kafka managed services or between a Strimzi-based Apache Kafka cluster and another managed Kafka cluster.
+  This can be part of a one-off migration project.
+  But it can also be part of a long-term hybrid cloud setup that combines different cloud providers and on-premise services.
+
+Adding support for custom client-side authentication would give these users a way to use the custom authentication mechanisms.
+It would allow them to improve their security (by using better mechanisms and better integration with other services).
+And it should also make it easier to configure their operands as it should be more obvious how to configure the authentication.
+
+## Proposal
+
+A new `type: custom` authentication type will be added to all places where we use client-side authentication today:
+* Kafka Connect
+* Strimzi HTTP Bridge
+* Kafka Mirror Maker 2
+
+The API for the new `custom` type client-side authentication would follow the existing server-side API.
+It will have a flag indicating whether the authentication mechanism uses SASL.
+And it will have a `config` section for adding additional configuration options such as `sasl.mechanism` or JAAS configuration.
+
+The `secrets` field that is present in the server-side `custom` authentication but is already deprecated will not be used in the client-side API.
+Users can use the _additional volumes_ feature to add custom secrets required by the authentication mechanism.
+
+Users can use configuration providers for values such as passwords to load them from files or environment variables without having them hardcoded and visible in the Strimzi custom resources.
+
+The `config` field will allow configuration of the fields starting with the following prefixes:
+* `ssl.keystore.`
+* `sasl.`
+
+These options are otherwise forbidden and cannot be set in the regular _configuration_ sections (i.e. in `.spec.config` or `.spec.clusters.config`).
+Other options which might be needed by the custom authentication mechanisms can be configured in the regular _configuration_ sections.
+That way, we make sure that users are able to configure the custom mechanisms, but cannot use it to override the other configuration options blocked by Strimzi (e.g. the REST API configuration for Connect) by mistake or intention.
+
+Use of TLS and the configuration of trusted certificates will continue to be done through the corresponding `tls` sections.
+It will not be configurable through the custom authentication mechanism.
+
```

Comment:
shall we say how invalid config is handled?

Reply:
I extended the proposal.